regparser/api_writer.py

```diff
@@ -123,16 +131,396 @@ def write(self, python_object):
+        all_offsets = []
+        all_replacements = []
+        # import pdb; pdb.set_trace()
+        for ident, layer in self.layers.items():
+            # print node.label_id()
+            if node.label_id() in layer:
+                # print 'applying layers'
+                replacements = layer[node.label_id()]
+                if ident == 'terms':
+                    offsets, repls = XMLWriteContent.apply_terms(text, replacements)
+                elif ident == 'paragraph-markers':
+                    offsets, repls = XMLWriteContent.apply_paragraph_markers(text, replacements)
+                elif ident == 'internal-citations':
+                    offsets, repls = XMLWriteContent.apply_internal_citations(text, replacements)
+                elif ident == 'definitions':
+                    offsets, repls = XMLWriteContent.apply_definitions(text, replacements)
+                elif ident == 'external-citations':
+                    offsets, repls = XMLWriteContent.apply_external_citations(text, replacements)
+                else:
+                    offsets = []
+                    repls = []
+
+                all_offsets.extend(offsets)
+                all_replacements.extend(repls)
+
+        if len(all_offsets) > 0 and len(all_replacements) > 0:
+
+            offsets_and_repls = zip(all_offsets, all_replacements)
+            offsets_and_repls = sorted(offsets_and_repls, key=lambda x: x[0][0])
+            all_offsets, all_replacements = zip(*offsets_and_repls)
+
+            text = interpolate_string(text, all_offsets, all_replacements).strip()
+
+        return text
+
 class Client:
     """A Client for writing regulation(s) and meta data."""
 
-    def __init__(self):
-        if settings.API_BASE:
-            self.writer_class = APIWriteContent
-        elif getattr(settings, 'GIT_OUTPUT_DIR', ''):
-            self.writer_class = GitWriteContent
+    def __init__(self, writer_type=None):
+        if writer_type is None:
+            if settings.API_BASE:
+                self.writer_class = APIWriteContent
+            elif getattr(settings, 'GIT_OUTPUT_DIR', ''):
+                self.writer_class = GitWriteContent
+            else:
+                self.writer_class = FSWriteContent
         else:
-            self.writer_class = FSWriteContent
+            if writer_type == 'API':
+                self.writer_class = APIWriteContent
+            elif writer_type == 'git':
+                self.writer_class = GitWriteContent
+            elif writer_type == 'XML':
+                self.writer_class = XMLWriteContent
+            elif writer_type == 'file':
+                self.writer_class = FSWriteContent
+            else:
+                raise ValueError('Unknown writer type specified!')
+
+    def reg_xml(self, label, doc_number, layers=None):
+        return self.writer_class("regulation/{}/{}.xml".format(label, doc_number), layers=layers)
+
+    def notice_xml(self, doc_number):
```

Comment:
Is there a reason, besides path, that these need to be separate from `regulation()` and `notice()` below? I think there should be a better way to combine that without adding XML-specifics to this class.